From c627b9c3c7fc7ce59a2696dd3012a08b18e6b1cb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jonas=20Dre=C3=9Fler?= <verdre@v0yd.nl>
Date: Mon, 14 Nov 2022 00:14:05 +0100
Subject: [PATCH] clutter/actor: Show on all stage-views when actors have no
 allocation

When a badly behaving ClutterActor implementation manages to invalidate
the allocation after the layout phase and before painting, we have no
idea where the actor should be painted without running the whole layout
machinery again.

For paint volumes in this case we pretend the actor covers the whole
stage and queue full-stage redraws. When updating stage-views, we're
also handling this case, but not in the most graceful way. Just like
with paint volumes, we should assume an actor without a valid allocation
is simply everywhere, so set priv->stage_views to all available stage
views in that case.

Related: https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/6054
---
 clutter/clutter/clutter-actor.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/clutter/clutter/clutter-actor.c b/clutter/clutter/clutter-actor.c
index 1dde8d9486..10a5e3203b 100644
--- a/clutter/clutter/clutter-actor.c
+++ b/clutter/clutter/clutter-actor.c
@@ -15507,6 +15507,9 @@ update_stage_views (ClutterActor *self)
   ClutterStage *stage;
   ClutterActorBox stage_paint_box;
   graphene_rect_t bounding_rect;
+  
+  stage = CLUTTER_STAGE (_clutter_actor_get_stage_internal (self));
+  g_return_if_fail (stage);
 
   old_stage_views = g_steal_pointer (&priv->stage_views);
 
@@ -15514,12 +15517,10 @@ update_stage_views (ClutterActor *self)
     {
       g_warning ("Can't update stage views actor %s is on because "
                  "!visible_paint_volume_valid.", _clutter_actor_get_debug_name (self));
+      priv->stage_views = g_list_copy (clutter_stage_peek_stage_views (stage));
       goto out;
     }
 
-  stage = CLUTTER_STAGE (_clutter_actor_get_stage_internal (self));
-  g_return_if_fail (stage);
-
   _clutter_paint_volume_get_stage_paint_box (&priv->visible_paint_volume,
                                              stage,
                                              &stage_paint_box);

-- 
GitLab

